src/par_iter/imp.rs

```diff
@@ -0,0 +1,79 @@
+use super::*;
+
+/// The `ParallelIterator` implementation interface.
+pub trait ParallelIteratorImpl: Sized {
+    type Item: Send;
+
+    /// Internal method used to define the behavior of this parallel
+    /// iterator. You should not need to call this directly.
+    fn drive_unindexed<C>(self, consumer: C) -> C::Result where C: UnindexedConsumer<Self::Item>;
+
+    /// Returns the number of items produced by this iterator, if known
+    /// statically. This can be used by consumers to trigger special fast
+    /// paths. Therefore, if `Some(_)` is returned, this iterator must only
+    /// use the (indexed) `Consumer` methods when driving a consumer, such
+    /// as `split_at()`. Calling `UnindexedConsumer::split_off_left()` or
+    /// other `UnindexedConsumer` methods -- or returning an inaccurate
+    /// value -- may result in panics.
+    ///
+    /// This is hidden & considered internal for now, until we decide
+    /// whether it makes sense for a public API.  Right now it is only used
+    /// to optimize `collect`  for want of true Rust specialization.
+    #[doc(hidden)]
+    fn opt_len(&mut self) -> Option<usize> {
+        None
+    }
+}
+
+impl<T> ParallelIterator for T where T: ParallelIteratorImpl {}
+
+
+/// A trait for parallel iterators items where the precise number of
+/// items is not known, but we can at least give an upper-bound. These
+/// sorts of iterators result from filtering.
+pub trait BoundedParallelIteratorImpl: ParallelIterator {
+    fn impl_upper_bound(&mut self) -> usize;
+
+    /// Internal method used to define the behavior of this parallel
+    /// iterator. You should not need to call this directly.
+    fn drive<'c, C: Consumer<Self::Item>>(self, consumer: C) -> C::Result;
+}
+
+impl<T> BoundedParallelIterator for T
+    where T: BoundedParallelIteratorImpl
+{
+    fn upper_bound(&mut self) -> usize {
+        self.impl_upper_bound()
+    }
```

Comment:
I don't love this indirection myself, but it seemed necessary if we want to keep `upper_bound` in the public API.  I wanted to avoid having to implement *two* bounded traits, instead of just requiring `*Impl` and everything is fine.  The `impl_` prefix is just to avoid calling ambiguity for anyone who has both traits in scope.  Ditto `ExactParallelIterator::len()` below.